modules/azure/client_factory.go

```
@@ -967,3 +1002,53 @@ func getBaseURI() (string, error) {
 	}
 	return baseURI, nil
 }
+
+// getArmAppContainersClientFactory gets an arm app containers client factory
+func getArmAppContainersClientFactory(subscriptionID string) (*armappcontainers.ClientFactory, error) {
+	targetSubscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	clientCloudConfig, err := getClientCloudConfig()
+	if err != nil {
+		return nil, err
+	}
+	cred, err := azidentity.NewDefaultAzureCredential(&azidentity.DefaultAzureCredentialOptions{
+		ClientOptions: azcore.ClientOptions{
+			Cloud: clientCloudConfig,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return armappcontainers.NewClientFactory(targetSubscriptionID, cred, &arm.ClientOptions{
+		ClientOptions: policy.ClientOptions{
+			Cloud: clientCloudConfig,
+		},
+	})
+}
+
+func getClientCloudConfig() (cloud.Configuration, error) {
+	envName := getDefaultEnvironmentName()
+	switch strings.ToUpper(envName) {
```

What if the envName doesn't match any known environment, it returns an error. You could consider logging the defaulting behavior or providing guidance on what values are expected.

---

I could find any good examples on how logging is done without having the `testing.T` at hand with `terratest.modules.logger`. Do you have any preferences or is it sufficient to just add enough context to the error returned?

---

Hmm can you just pass in the testing.T variable along and share the context that weay?

---

That would require all the funtions XxxxE like ManagedEnvironmentExistsE to break the pattern from all other azure test functions and have t testing.T as an argument. My initial thought behind just erroring out is that the value defaults to AzurePublicCloud and if the user has set the env var to a invalid value the test should fail fast and return an error with the reason. If my concerns around the different function pattern and fail fast with error is wrong I'm more than happy to make those changes

---

That's a valid point. We could look into better logging pattern in a separate issue.